R/backend-mssql.R

```diff
@@ -150,8 +150,27 @@
           sql_expr(DATEPART(QUARTER, !!x))
         }
       },
+    )
 
-    ),
+  if (mssql_version(con) >= "11.0") { # MSSQL 2012
+    mssql_scalar <- sql_translator(
+      .parent = mssql_scalar,
+      as.logical = sql_try_cast("BIT"),
+      as.Date = sql_try_cast("DATE"),
+      as.POSIXct = sql_try_cast("TIMESTAMP"),
+      as.numeric = sql_try_cast("FLOAT"),
+      as.double = sql_try_cast("FLOAT"),
+      as.integer = sql_try_cast("NUMERIC"),
+      # in MSSQL, NUMERIC converts to integer
```

Comment:
I wonder what's the reasoning behind this. When I do this, I get a `numeric` via FreeTDS:

``` r
con <- DBI::dbConnect(
  odbc::odbc(),
  "mssql-test",
  uid = "kirill", pwd = keyring::key_get("mssql", "kirill")
)

library(tidyverse)

tbl <-
  tibble(a = 1L) %>%
  copy_to(con, ., name = "tbl_a") %>%
  mutate(b = as.integer(a))
#> Created a temporary table named #tbl_a

tbl
#> # Source:   lazy query [?? x 2]
#> # Database: Microsoft SQL Server 15.00.4073[@tp/test]
#>       a     b
#>   <int> <dbl>
#> 1     1     1
dbplyr::sql_render(tbl)
#> <SQL> SELECT "a", TRY_CAST("a" AS NUMERIC) AS "b"
#> FROM "#tbl_a"
```

<sup>Created on 2020-11-16 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0)</sup>

Reply:
In https://github.com/DavidPatShuiFong/dbplyr/commit/4933037fcdea29606ac29badefecee9e753b6eef I describe how I found that "in MSSQL casting to NUMERIC casts to integer, and is more flexible/graceful ... [NUMERIC] returns NA if unable to convert individual rows. Whereas casting to  INTEGER/INT results in the entire column becoming NA if a single row cannot be converted to integer."

Reply:
Yeah, a double cast might work -- would still all rows be converted to `NULL` if the numbers are too large or too small?

Reply:
yes, double casting works! and doesn't drop the entire column if a single row has an invalid value.

Happy to know if there is a more elegant solution, but here is what I did (from https://github.com/tidyverse/dbplyr/compare/master...DavidPatShuiFong:feature/doubletrycastMSSQL )

changed `try_cast` (`translate_sql_helpers.R`) to accept a second optional parameter to do a second casting

```r
#' @rdname sql_variant
#' @export
sql_try_cast <- function(type1, type2 = NULL) {
  type1 <- sql(type1)
  if (is.null(type2)) {
    function(x) {
      sql_expr(try_cast(!!x %as% !!type1))
      # try_cast available in MSSQL 2012+
    }
  } else {
    type2 <- sql(type2)
    # if type2 is defined, then double-cast ('type1' then 'type2')
    function(x) {
      sql_expr(try_cast(try_cast(!!x %as% !!type1) %as% !!type2))
      # try_cast available in MSSQL 2012+
    }
  }
}
```

Change the casting of `as.integer` in `backend-mssql.R` for `mssql_version` >= 11.0

```r
 as.integer = sql_try_cast("NUMERIC", "INT"),
```

For the example I've given (I couldn't set up your example to work, perhaps because I am using an old version of `dplyr`?)

```r
> x <- con %>% dplyr::tbl("##nchild") %>% dplyr::mutate(n = as.integer(n))
> dbplyr::sql_render(x)
<SQL> SELECT "Year", "Country", TRY_CAST(TRY_CAST("n" AS NUMERIC) AS INT) AS "n"
FROM "##nchild"
> x
# Source:   lazy query [?? x 3]
# Database: Microsoft SQL Server 12.00.2000[bpsrawdata@DESKTOP-BH8LGMO\BPSINSTANCE/bpssamples]
   Year                Country     n
   <chr>               <chr>   <int>
 1 1970-01-01 10:32:50 China      NA
 2 xyz                 China       4
 3 1970-01-01 10:33:00 China       3
 4 1970-01-01 10:33:05 China       3
 5 1970-01-01 10:33:10 China       2
 6 1970-01-01 10:33:15 China       2
 7 1970-01-01 10:33:20 China       2
 8 1970-01-01 10:33:25 China       2
 9 1970-01-01 10:33:30 China       2
10 1970-01-01 10:32:50 Taiwan      4
# ... with more rows
```

Reply:
Thanks, this looks good. I don't know an easier solution off the top of my head.

Reply:
> 
> 
> Thanks, this looks good. I don't know an easier solution off the top of my head.

Thanks @krlmlr . After a bit more testing on the 'production code' which uses `dbplyr`, should I open a new issue and submit the fix as a pull request?

Reply:
Thanks. A PR referring to this conversation should be enough.